### PR TITLE
Remove readOrNewId and add "verify 3PC" endpoint to operator API

### PR DESCRIPTION
**Pull request by OlivierChirouze**
Add a description of "verify 3PC" endpoint that is needed to check that 3PC are supported (see https://github.com/criteo/addressable-network-proposals/pull/27)

### CODE REVIEW

File: mvp-spec/operator-api.md [hunk 1]
```diff
@@ -16,13 +16,13 @@ These endpoints need to support both the "3PC" and "no 3PC" contexts.
...
-| Read data                               | -                       | List of IDs (if any).<br>List of preferences (if any) | If ID 🍪 exist, return it.<br>If preferences 🍪 exist, return them                                                                                                                        | -                                                                                                                                                                                                                                                                                       | GET /v1/json/read           | GET /v1/redirect/read           |
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:21]: I am ok for the cookies emoticons in the schemas but I am more skeptical about in the written documentation. What about using plain English ?

File: mvp-spec/operator-api.md [hunk 2]
```diff
@@ -16,13 +16,13 @@ These endpoints need to support both the "3PC" and "no 3PC" contexts.
...
+| Read data if exists, or return a new Id | -                       | List of IDs.<br>List of preferences (if any)<br>newly generated ID, if any | If ID 🍪 exist, return it.<br><br>Otherwise: create a new ID, sign it and return it. In this case, the **new ID is not stored in a cookie yet**..<br>If preferences 🍪 exist, return them<br><br>Note: when called in REST mode, a special short-life "3PC" cookie is also set | Why don't we save the newly generated ID?<br>Because haven't yet received consent. The new ID will be saved as cookie, along with preferences, when it is provided via the "Write data" endpoint ⬇️                                                                                     | GET /v1/json/read      | GET /v1/redirect/read  |
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:21]: Either above of the table or within the first column, it is good to remind that the output is conditioned by the session of the user. Therefore, two users has different output when reaching the same endpoint.

File: mvp-spec/operator-api.md [hunk 3]
```diff
@@ -16,13 +16,13 @@ These endpoints need to support both the "3PC" and "no 3PC" contexts.
 
 In practice, this will translate into endpoints available under different root paths. Example paths are specified in the last column of the table.
 
-| Endpoint                                | Input*                  | Output                                                | Description                                                                                                                                                                               | Notes                                                                                                                                                                                                                                                                                   | Rest                        | Redirect                        |
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:19]: Here, "REST" is inaccurate. In REST, GET is safe (i.e read-only) whereas we potentially create a resource in our case. 

Possible change:
1. *Direct*. 
   * Pro: Symmetry with "Redirect". 
   * Cons: Subject to typo/error because the two words are closed.
2. *On Site*
   * Pro: Explicite on how it is used
   * Cons: May not match all the case.
3. *Direct Data*
   * Pro: agnostic of the format, symmetry

What do you think ? :-)

File: mvp-spec/operator-api.md [hunk 4]
```diff
@@ -16,13 +16,13 @@ These endpoints need to support both the "3PC" and "no 3PC" contexts.
...
-| Write data.<br>and return written data  | main ID.<br>preferences | List of IDs.<br>List of preferences                   | Read provided ID.<br>Verify provided signatures.<br>Write ID 🍪.<br>Write preferences 🍪                                                                                                  | ID is mandatory input for the "first visit" use case so is considered always mandatory, for consistency.<br>Data is written because it will be used for confirmation.                                                                                                                   | POST /v1/json/write         | GET /v1/redirect/write          |
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:23]: We mention a mandatory input, here. However, it doesn't appear in the detail of the API.
Do we want to be precise here ? What do you think about adding a payload for the POST and a URL param (or a path) for the GET? Or maybe in URL for both for consistency?

File: mvp-spec/operator-api.md [hunk 5]
```diff
@@ -16,13 +16,13 @@ These endpoints need to support both the "3PC" and "no 3PC" contexts.
...
+| Get new ID                              | -                       | main ID                                                                    | create a new ID and sign it.<br>do not store a cookie.<br>return ID                                                                                                                                                                                                            | Cookie is not saved at this stage because we show the new value to the user before they validate it (and then it is saved, using the "Write data" endpoint ⬆️ ).<br>Since there is no cookie to read or write, no redirect version is needed, it can be made in JS with our without 3PC | GET /v1/json/newId     | Not available          |
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:23]: /json/ appear on every "REST" API. Do we expect to have another format? Does it make sense to remote it?

File: mvp-spec/operator-api.md [hunk 6]
```diff
@@ -16,13 +16,13 @@ These endpoints need to support both the "3PC" and "no 3PC" contexts.
...
+| Verify 3PC                              | -                       | boolean                                                                    | Attempts to read the short-life cookie that was set by the "Read" endpoint.<br>Return true if found, false if not.                                                                                                                                                             | The goal is to confirm that 3PC are supported. If the cookie set at previous step can be read, it means 3PC are supported.<br>See [website-design](./website-design.md) for details.<br><br>Note this endpoint doesn't require any signature.                                           | GET /v1/json/verify3PC | Not available          |
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:24]: If you want to be closer to a REST API. Consider the 3PC as a resource that can be here or not. Therefore, "verify" in the path `/v1/json/verify3PC` is not really necessary.

File: mvp-spec/operator-api.md [hunk 7]
```diff
@@ -213,93 +213,34 @@ cat response-operatorO.json body-id-and-preferences.json | npx json --merge -e '
...
-      {
-        "version": 1,
-        "type": "prebid_id",
-        "value": "7435313e-caee-4889-8ad7-0acd0114ae3c",
-        "source": {
-          "domain": "operator0.com",
-          "timestamp": 1639643110,
-          "signature": "prebid_id_signature_xyz12345"
-        }
+    "newIdentifier": {
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:216]: I would keep the same structure with a list of "identifiers" and avoiding "newIdentifier". If you want to express the creation of the resource, maybe we can use the standard of HTTP by returning a 201 status instead of a 200.

File: mvp-spec/operator-api.md [hunk 8]
```diff
@@ -213,93 +213,34 @@ cat response-operatorO.json body-id-and-preferences.json | npx json --merge -e '
...
 
 ```
 sender + '\u2063' +
 receiver + '\u2063' +
-preferences.source.signature + '\u2063' +
-identifiers[0].source.signature + '\u2063' +
-identifiers[1].source.signature + '\u2063' +
-...
-identifiers[n].source.signature + '\u2063' +
+newIdentifier.source.signature + '\u2063' +
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:237]: If you accept the previous comment about using "identifiers", don't forget to update it here.

File: mvp-spec/operator-api.md [hunk 9]
```diff
@@ -451,18 +392,66 @@ cat response-operatorO.json body-id.json | npx json --merge
...
+```
+sender + '\u2063' +
+receiver + '\u2063' +
+newIdentifier.source.signature + '\u2063' +
+timestamp
+```
+
+### GET /v1/json/verify3PC
+
+This endpoint is very "technical" and used only when a first attempt to read Prebid cookies via REST failed.
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:422]: I guess that at this level everything is technical. :D 
Put simply: "This endpoint is used when a first attempt to read..."

File: mvp-spec/operator-api.md [hunk 10]
```diff
@@ -451,18 +392,66 @@ cat response-operatorO.json body-id.json | npx json --merge
...
+```
+
+**No signature** is required.
+
+#### Response
+
+- 3PC supported (test cookie was found)
+  
+  ```json
+  true
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:444]: For a matter of consistency, use a JSON and not a plain value here. It is also counter-intuitive because json is in the path).
You can also use a HTTP status to express result. Example: 200 for "true case", 404 or 406 for "false case".

File: mvp-spec/operator-api.md (hunk 7 above)
Comment (OlivierChirouze) [reply mvp-spec/operator-api.md:223]: After a second thought, I'm not so sure that this would be appropriate: 201 would mean that a resource was created, but in this case the ID is "generated" but it's not stored as a cookie. So I think it's not appropriate to consider this as a resource creation.
In fact, it would be probably more appropriate to return a 404 but I'm concerned about processing the response content with 404 HTTP code.

Again, if we consider this call is always made by "our own" JS library, then processing a 404 response is probably fine, and more RESTfull. (and the fetch API [supports it](https://stackoverflow.com/questions/39297345/fetch-resolves-even-if-404))

File: mvp-spec/operator-api.md (hunk 3 above)
Comment (RomainLofaso) [reply mvp-spec/operator-api.md:19]: I think that you have improved your API with your last commit and It is a good thing. It is more idiomatic now. 

If I check this [maturity model](https://en.wikipedia.org/wiki/Richardson_Maturity_Model), I guess that we are between level 1 and level 2 of REST. So it may not be RESTful. It is not really clear to me if our API is resource-oriented. The `GET /v1/id-prefs` isn't idempotent. However, do we really care? We don't need to be RESTful. REST doesn't exist for itself. It is a set of principles to ease scalability, uniformity, and layered architectures. You can have those properties without being RESTful.

So regarding this, I am still tempted to remove the REST word from the header of the table.

File: mvp-spec/operator-api.md (hunk 7 above)
Comment (RomainLofaso) [reply mvp-spec/operator-api.md]: Interesting perspective. The 404 doesn't follow the principle of least surprise, here, IMHO.

I am wondering if it is not due to the nature of the ID and the flow of data. I think that we have an implicit state in the ID. First, it is transient and then it can be stored. Maybe we should express this state in the resource. What do you think ?

Example:

```json
{
  "sender": "operatorO.com",
  "timestamp": 1639059692793,
  "signature": "message_signature_xyz1234",
  "body": {
    "identifiers": [
      {
        "version": 1,
        "state": "transient",
        "type": "prebid_id",
        "value": "560cead0-eed5-4d3f-a308-b818b4827979",
        "source": {
          "domain": "operator0.com",
          "timestamp": 1639643110,
          "signature": "prebid_id_signature_xyz12345"
        }
      }
    ]
  }
}
```
The downside of this approach is that we have an extra field that may not be necessary once the identifier is stored.

Comment (RomainLofaso) [reply mvp-spec/operator-api.md]: Side note concerning your first thought: It can be also error-prone to have similar names around the same place of the payload. Indeed, it is not easy to catch while reading a documentation. It makes me think about some guidances from Clean Code: we should avoid long names for variables and functions with a delta of a few letters.

File: mvp-spec/versioning.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# Versioning
```
Comment (RomainLofaso) [inline mvp-spec/versioning.md:1]: IMO, this document in its current state is a bit premature and should be done in another PR.

Comment (OlivierChirouze) [reply mvp-spec/versioning.md:1]: it has nothing to do in this PR... fail! 🤦‍♂️
Sorry about that.

File: mvp-spec/operator-api.md (hunk 3 above)
Comment (OlivierChirouze) [reply mvp-spec/operator-api.md:19]: As you say, REST doesn't even really exist so I bet most of the APIs in the wild that call themselves "REST" aren't RESTfull 😁
In that sense, I don't think it's a big deal to leave the REST keyword in this table. Specially, I'm concerned renaming it "direct" or "on site" would bring more confusion than clarity, don't you think?
Unless we find a name that we agree on, I would leave REST for now

Comment (OlivierChirouze) [reply mvp-spec/operator-api.md:19]: After our last discussion, I will leave REST in this table, but add a quick "disclaimer".

File: mvp-spec/operator-api.md [hunk 12]
```diff
@@ -265,26 +230,28 @@ cat response-operatorO.json body-id-and-preferences.json | npx json --merge -e '
...
         "type": "prebid_id",
-        "value": "7435313e-caee-4889-8ad7-0acd0114ae3c",
+        "value": "560cead0-eed5-4d3f-a308-b818b4827979",
         "source": {
           "domain": "operator0.com",
           "timestamp": 1639643110,
           "signature": "prebid_id_signature_xyz12345"
-        }
+        },
+        "persisted": false
```
Comment (RomainLofaso) [inline mvp-spec/operator-api.md:243]: Not clear to me. It seems that `persisted` is in a middle of an array.
